Example set for using OneAPI with CMake (#1721)
oneapi-src/oneAPI-samples: Templates/cmake/README.md

# CMake Template Examples for OneAPI

The CMake projects in this code repository demonstrate use of CMake to build
simple programs with OneAPI compilers for various common scenarios.

| Area                      | Description
|:---                       |:---
| What you will learn       | How to use CMake to build projects with OneAPI compilers.
| Time to complete          | 10 minutes
| Category                  | Tutorial

## Purpose

The CMake projects in this code repository demonstrate use of CMake to build
simple programs with OneAPI Compilers for various common scenarios. Examples are divided
-into directories for C, C++, and Fortran. SYCL examples are included at the
top level. SYCL examples are implemented entirely in C++.

## Prerequisites

| Optimized for             | Description
|:---                       |:---
| OS                        | Windows, Linux
| Hardware                  | N/A
| Software                  | CMake, OneAPI Compilers

## Key Implementation Details

> Each language specific directory are contains few sample examples such as OpenMP, IPO.

Author: done

The examples in this directory are structured as a collection of independent
projects, rather as a large single project. This way, any example can be copied
into a separate subdirectory, compiled, run, and used as the basis for a new project.

The top level CMakeLists.txt includes projects in all the sub-directories.  To
build all of the examples, create a build directory and generate the project as
usual.  
 

> For example in Linux,

Author: done

    $ mkdir build
    $ cd build

>     $ CC=icx CXX=icpx FC=ifx cmake ..

Author: Yes, We can do that but that requires changing CMakeLists.txt by user and thought command line is easier. Another reason is that user will be forced to think/ensure the availability of oneAPI compilers. I can change it as you suggested if these are not an issue..

>     $ CC=icx CXX=icpx FC=ifx cmake ..

Reviewer: My thoughts are that: This is a OneAPI compiler cmake tutorial, so I would assume that these compilers are available to the user. Maybe it would be worth to only set the compiler that is required by the current project being compiled (so that cmake does not check if FC is working when the project is compiled with the SYCL compiler.
I find the error message ok when cmake does not find the compiler through: set(CMAKE_CXX_COMPILER icpx)

>     $ CC=icx CXX=icpx FC=ifx cmake ..

Author: Ok, another thought I just realized is that we support a combination of drivers in oneAPI
for eg:
Just for C++, 
In windows:
     native style: icx
     GNU style: icpx
In Linux:
     native style: icpx
     windows style: icx-cl
We have similar combination with other C/SYCL as well

Do we still want to have it added to CMakeLists.txt instead of commandline?

>     $ CC=icx CXX=icpx FC=ifx cmake ..

Reviewer: My feeling is that:
1/ We want users to use icpx on Linux and icx-cl on Windows. Let's make it clear by having it directly in the cmake file. The only reason one would want to change the compiler being used is if they want to use specific compiler flags. In this situation, they'll need to modify the cmake anyway, so they can change the compiler if needed.
2/ ipcx on Windows does not really work (through cmake I mean), I suspect icx-cl has similar issues on Linux.
3/ The flags you use in your current cmake would work for all these combinations? 

For all these reasons, I'd say we should set icpx on Linux and icx-cl on Windows (and test both), what are your thoughts?

>     $ CC=icx CXX=icpx FC=ifx cmake ..

Author: Ok then, I will update the scripts and document

>     $ cmake --build . -j

Author: I will remove

>     $ cmake --build . -j

Author: done

For example in Windows

    $ mkdir build
    $ cd build

>     $ CC=icx CXX=icx FC=ifx cmake -GNinja ..

Author: All our ICX Windows support 4so far is for Ninja only. The "NMake Makefiles" isn't added/maintained by us, atleast for now.

>     $ CC=icx CXX=icx FC=ifx cmake -GNinja ..

Reviewer: Interesting, so it happens to work, but all the FPGA samples rely on it, and we do test all the samples with it at each release, without any issues.

    $ cmake --build . -j

> >**Note**: CMake support for GNU style OneAPI driver `icpx` on Windows is not yet available until CMake 3.25 and such support will be added in future releases of CMake.

Author: done

>**Note**: For comprehensive information about oneAPI programming, see the *[Intel® oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide)*. (Use search or the table of contents to find relevant information quickly.)
>**Note**: For comprehensive information about CMake with OneAPI Compilers, see the *[Intel® oneAPI DPC++/C++ Compiler Developer Guide and Reference](https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference)*. (Use search or the table of contents to find relevant information quickly. OR, Navigate to Compiler Setup -> Use the Command Line -> Use the Cmake with the compiler)

## Set Environment Variables

When working with the command-line interface (CLI), you should configure the oneAPI toolkits using environment variables. Set up your CLI environment by sourcing the `setvars` script every time you open a new terminal window. This practice ensures that your compiler, libraries, and tools are ready for development.

> Ensure the availability of latest CMake binaries.

Author: done

## Build the Cmake example Sample

> **Note**: If you have not already done so, set up your CLI
> environment by sourcing  the `setvars` script in the root of your oneAPI installation.
>
> Linux*:
> - For system wide installations: `. /opt/intel/oneapi/setvars.sh`
> - For private installations: ` . ~/intel/oneapi/setvars.sh`
> - For non-POSIX shells, like csh, use the following command: `bash -c 'source <install-dir>/setvars.sh ; exec csh'`
>
> Windows*:
> - `C:\Program Files (x86)\Intel\oneAPI\setvars.bat`
> - Windows PowerShell*, use the following command: `cmd.exe "/K" '"C:\Program Files (x86)\Intel\oneAPI\setvars.bat" && powershell'`
>

> > For more information on configuring environment variables, see *[Use the setvars Script with Linux* or macOS*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html)* or *[Use the setvars Script with Windows*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html)*.

Author: done

### On Linux*

1. Change to the sample directory.
2. Build the program.
   ```
    $ mkdir build
    $ cd build
    $ CC=icx CXX=icpx FC=ifx cmake ..
    $ cmake --build . -j
   ```

### On Windows*

1. Change to the sample directory.
2. Build the program.
   ```
    $ mkdir build
    $ cd build
    $ CC=icx CXX=icx FC=ifx cmake -GNinja ..
    $ cmake --build . -j
   ```

> >**Note**: Currently, only Ninja generators are supported in the Windows. A Visual Studio Generator may be supported with OneAPI may be supported in the future CMake releases.

Reviewer: ```suggestion
>**Note**: Currently, only Ninja generators are supported on Windows.
```

I don't think that is true as all the FPGA samples use "NMake Makefiles" as a generator

Author: I meant the support we added/maintained.

Author: done

> CMake + Intel&reg; oneAPI Examples

Reviewer: The Readme would need to be modified to match the existing template

Author: Oops, sorry, I made modifications but in a local copy, missed that out. I will update. Thanks for pointing.
